pytorch_lightning/loops/fit_loop.py

```diff
@@ -142,31 +157,23 @@ def done(self) -> bool:
         """Evaluates when to leave the loop."""
         # TODO(@awaelchli): Move track steps inside training loop and move part of these condition inside training loop
         stop_steps = _is_max_limit_reached(self.global_step, self.max_steps)
-        stop_epochs = _is_max_limit_reached(self.epoch_progress.current.completed, self.max_epochs)
-
-        should_stop = False
-        if self.trainer.should_stop:
-            # early stopping
-            met_min_epochs = self.epoch_progress.current.completed >= self.min_epochs if self.min_epochs else True
-            met_min_steps = self.global_step >= self.min_steps if self.min_steps else True
-            if met_min_epochs and met_min_steps:
-                should_stop = True
-            else:
+        stop_epochs = _is_max_limit_reached(self.epoch_progress.current.processed, self.max_epochs)
+
+        if self.trainer.should_stop and self.min_epochs:
```

Comment:
I understand that is now a limitation. But is it a bug of a feature?

For proper loop customization we need to respect the hierarchy imposed by the loops. The way they are, epochs go before steps so they should always be prioritized when both are set.

Reply:
> The way they are, epochs go before steps so they should always be prioritized when both are set.

I disagree. There is no reason to believe that "epochs" are more important than "steps". A setting like `Trainer(min_epochs=1, min_steps=50)` would be contradictory if we stopped at 1 epoch without reaching 50 steps. The arguments in the Trainer mean, if I request `Trainer(x=A, y=B, z=C)` I want features A AND B AND C together, not A OR B OR C.

Reply:
I don't think we can do that, because the intersection depends on the structure of the loops.

When `fit_loop.done is True` we would need to ask the children loops, but to do that they need to run their `reset` hook which is only called in their `run`.

We cannot know if the user wants the intersection or the union.